### DISCUSSION TITLE
Allow the re-use of internal parts to simplify customization

### DISCUSSION THREAD

**Discussion by jrmyio** [nextauthjs/next-auth#4069] [Ideas]
When trying to extend or modify `next-auth` default behaviors I ran into several problems where the core does not seem to be flexible enough and there seem to be no way to re-use some of its parts. In a PR (https://github.com/nextauthjs/next-auth/pull/4036) I tried to add more flexibility but as @balazsorban44 mentioned adding more configuration is probably not the right way to do this.

The docs recommend using a high-level method: [initialization#advanced-initialization](https://next-auth.js.org/configuration/initialization#advanced-initialization). However, this code is ran _BEFORE_ next-auth does its internal magic, thus you miss out on some of the internal implementations and framework agnostic abstractions.

In an attempt to build things on top of `next-auth` I ran into the following problems with its internals:

### Problem 1: Redirect responses instead of errors
When consuming next-auth as a Rest API one would want JSON responses that describes the errors with a message and/or code. However, the internal route handling currently (only) returns 'redirects' with error messages embedded.

Example:
https://github.com/nextauthjs/next-auth/blob/3d0c68d9e00c707f254d17a511c54af4ebb9eeb0/packages/next-auth/src/core/routes/signin.ts#L70-L75

#### Possible solution:
Introduce an `error` object (possibly with an Error.type = enum) on [OutgoingResponse](https://github.com/nextauthjs/next-auth/blob/a72f1b6d21da89b4223542006c865862635c027b/packages/next-auth/src/core/index.ts#L31-L39) that can then be used in favor of the `redirect` for rest APIs.

### Problem 2: Initialization function of next-auth configuration is not exported
The initialization of the next-auth configuration is being done pretty deep (just before the route handling): https://github.com/nextauthjs/next-auth/blob/a72f1b6d21da89b4223542006c865862635c027b/packages/next-auth/src/core/index.ts#L72-L81 But when making custom routes (as explained in [initialization#advanced-initialization](https://next-auth.js.org/configuration/initialization#advanced-initialization), one does not have access to an asserted and initialization next-auth configuration yet (just the `userOptions`).

#### Possible solution:
Export the [init function](https://github.com/nextauthjs/next-auth/blob/a72f1b6d21da89b4223542006c865862635c027b/packages/next-auth/src/core/init.ts) and [assetConfig](https://github.com/nextauthjs/next-auth/blob/a72f1b6d21da89b4223542006c865862635c027b/packages/next-auth/src/core/lib/assert.ts) in `next-auth` built package.

**Note:** It might also be good to move any config related code (init, assert, related errors) to different folder, for example: [/packages/next-auth/src/core/config](https://github.com/nextauthjs/next-auth/tree/a72f1b6d21da89b4223542006c865862635c027b/packages/next-auth/src/core).

### Problem 3: Request/Response transformation to framework agnostic code not reusable
The IncomingRequest:
https://github.com/nextauthjs/next-auth/blob/a72f1b6d21da89b4223542006c865862635c027b/packages/next-auth/src/core/index.ts#L13-L24  
and OutgoingResponse:
https://github.com/nextauthjs/next-auth/blob/a72f1b6d21da89b4223542006c865862635c027b/packages/next-auth/src/core/index.ts#L31-L39 are nice abstractions to create framework agnostic request/response.

However, since the transformation from and to these types are done internally [here](https://github.com/nextauthjs/next-auth/blob/main/packages/next-auth/src/next/index.ts#L26-L60), you have to re-implement
them when you want to extend next-auth (in next.js) as proposed in [initialization#advanced-initialization](https://next-auth.js.org/configuration/initialization#advanced-initialization). 

**Related question:** When devs want create additional framework agnostic functionality do we expect them to implement on these IncomingRequest/OutgoingResponse abstractions? 

#### Possible solution:
Extract the transformation in a seperate function that can be called manually when required.

### Notes 
If the internals are not set in stone yet, and as a result exporting internal parts of `next-auth` is not an option because of BC, I suggest the use of `next-auth/internals`  instead of `next-auth` until the types are fleshed out while warning people about `next-auth/internals` being subject to change with each next-auth version.

### Contributing 🙌🏽
Yes, I am willing to help implement this feature in a PR.

**Comment by longlostnick** [comment]
@jrmyio appreciate you listing this all out! Did you ever get any of this figured out? My biggest problem right now is #2. Even having just this ability would unlock all sorts of flexibility.
